### PR TITLE
change: Implement test mechanism for Pipeline variables

### PR DESCRIPTION
**Pull request by qidewenwhen**
*Description of changes:* Implement test mechanism for Pipeline variables to check if they are compatible with downstream classes

*Testing done:* N/A

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/unit/sagemaker/workflow/test_mechanism/test_code/test_pipeline_var_compatibility_template.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+            step_dsl (str): The step definition retrieved from the pipeline definition DSL.
+            step_dsl_obj (List[dict]): The json load object of the step definition.
+        """
+        # TODO: remove the following hard code assertion once recursive assignment is added
+        if issubclass(self.clazz, Processor):
+            if param_with_none != "network_config":
+                assert '{"Get": "Parameters.nw_cfg_subnets"}' in step_dsl
+                assert '{"Get": "Parameters.nw_cfg_security_group_ids"}' in step_dsl
+                assert '{"Get": "Parameters.nw_cfg_enable_nw_isolation"}' in step_dsl
+            if issubclass(self.clazz, SageMakerClarifyProcessor):
```
Comment (navaj0) [inline tests/unit/sagemaker/workflow/test_mechanism/test_code/test_pipeline_var_compatibility_template.py:386]: Is this even supported?

File: tests/unit/sagemaker/workflow/test_mechanism/test_code/test_pipeline_var_compatibility_template.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+                    assert '{"Get": "Parameters.proc_input_s3_data_type"}' in step_dsl
+                    assert '{"Get": "Parameters.proc_input_app_managed"}' in step_dsl
+        elif issubclass(self.clazz, EstimatorBase):
+            if (
+                param_with_none != "instance_groups"
+                and self.default_args[CLAZZ_ARGS]["instance_groups"]
+            ):
+                assert '{"Get": "Parameters.instance_group_name"}' in step_dsl
+                assert '{"Get": "Parameters.instance_group_instance_count"}' in step_dsl
+            if issubclass(self.clazz, AmazonAlgorithmEstimatorBase):
```
Comment (navaj0) [inline tests/unit/sagemaker/workflow/test_mechanism/test_code/test_pipeline_var_compatibility_template.py:407]: I'm OK if we skip those AmazonAlgorithmEstimators

**Review by navaj0 — COMMENTED** [review]
update the name of the test folders

from 
```
test_mechanism/
  test_code/
  test_entries/
```
to

```
test_pipeline_variable_compatibility/
  template/
  test_components/
```

**Comment by mufaddal-rohawala** [comment]
@qidewenwhen is this PR relevant, still in draft?

**Comment by qidewenwhen** [comment]
> @qidewenwhen is this PR relevant, still in draft?

Hey @mufaddal-rohawala, the test mechanism introduced in this PR has been deprioritized. And some updates are required on top of this PR considering the Pipeline cache improvement we released in 10/2022. I'll close this PR at this point and reopen in the future if we resume the work.

**Pull request closed without merging**